File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    
    def dfs(numbers, step, total, target):
        if step == len(numbers):
            if total == target:
                return 1
            else:
                return 0
        res = 0
        res += dfs(numbers, step + 1, total + numbers[step], target) # 시작이 +인 서브트리에서 타겟넘버 만든 횟수
        res += dfs(numbers, step + 1, total - numbers[step], target) # 시작이 -인 서브트리에서 타겟넘버 만든 횟수
        return res
    
    answer = dfs(numbers, 0, 0, target)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int N;
    static int answer = 0;
    static int[] arr;
    static int T;
    public int solution(int[] numbers, int target) {
        // 매번 2가지 경우의 수 현재 수에 다음 수를 더할지 뺄지
        N = numbers.length;
        arr = numbers;
        T = target;
        
        dfs(1, numbers[0]);
        dfs(1, -numbers[0]);
        return answer;
    }
    
    public static void dfs(int idx, int sum) {
        if (idx == N) {
            if (sum == T) answer++;
            return;
        }
        dfs(idx+1, sum+arr[idx]);
        dfs(idx+1, sum-arr[idx]);
    }
}